glib/src/subclass/types.rs

```diff
@@ -233,23 +234,33 @@ impl TypeData {
         self.parent_class
     }
 
-    /// Returns a pointer to the interface implementation specific data.
+    /// Returns a pointer to the class implementation specific data.
     ///
-    /// This is used for interface implementations to store additional data.
-    pub fn get_interface_data(&self, type_: ffi::GType) -> ffi::gpointer {
+    /// This is used for class implementations to store additional data.
+    pub fn get_class_data(&self, type_: ffi::GType) -> ffi::gpointer {
         unsafe {
-            if self.interface_data.is_null() {
+            if self.class_data.is_null() {
                 return ptr::null_mut();
             }
 
-            for &(t, p) in &(*self.interface_data) {
-                if t == type_ {
-                    return p;
-                }
-            }
+            *(*self.class_data).get(&type_).unwrap_or(&ptr::null_mut())
+        }
+    }
 
-            ptr::null_mut()
+    pub unsafe fn set(&mut self, type_: ffi::GType, data: ffi::gpointer) {
+        if self.class_data.is_null() {
+            self.class_data = Box::into_raw(Box::new(HashMap::new()));
         }
+        if self.class_data.as_ref().unwrap().get(&type_).is_some() {
+            panic!("The class_data already contains a key for {}", type_);
+        }
+        (*(self.class_data as *mut HashMap<ffi::GType, ffi::gpointer>)).insert(type_, data);
+    }
+
+    pub unsafe fn get_mut(&mut self, type_: ffi::GType) -> ffi::gpointer {
```

Comment:
```suggestion
    pub unsafe fn get_class_data_mut(&mut self, type_: ffi::GType) -> ffi::gpointer {
```

And all as above for the normal getter